### PR TITLE
WIP: working dist for global and regular for browserify

### PR DESCRIPTION
**Pull request by gkatsev**
This is a WIP and would need to be cleaned up a bit (a lot (like npm scripts and removing `lib` and `dist` from the branch)).
This is more of a proof of concept for providing a `dist/bundle.js` file that assumes `video.js` is in the global scope but still allows users to require and build this plugin directly with browserify.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -58,12 +62,6 @@
     "video.js": "^5.0.0-rc.90",
     "watchify": "^3.4.0"
   },
-  "browserify": {
```
Comment (gkatsev) [inline package.json]: We no longer need a default transform for either the shim or babel.

File: package.json [hunk 2]
```diff
@@ -16,7 +16,10 @@
     "build-tests": "mkdir -p dist/ && browserify test/karma-qunit-shim.js test/*-test.js -u test/karma.conf.js -s videojs-dock -d -o dist/tests.js",
     "watch-tests": "mkdir -p dist/ && watchify test/karma-qunit-shim.js test/*-test.js -u test/karma.conf.js -s videojs-dock -v -d -o dist/tests.js",
     "test": "jshint src/**.js test/**.js index.js && npm run build-tests && karma start test/karma.conf.js --single-run",
-    "prepublish": "npm run build",
+    "predist": "npm run babel",
+    "dist": "mkdir -p dist/ && browserify -t browserify-shim . -s videojs-dock -o dist/bundle.js",
+    "babel": "babel src -d lib",
```
Comment (gkatsev) [inline package.json]: We want to run babel to compile files in `src` to `lib` from ES6 to ES5.

File: package.json [hunk 3]
```diff
@@ -16,7 +16,10 @@
     "build-tests": "mkdir -p dist/ && browserify test/karma-qunit-shim.js test/*-test.js -u test/karma.conf.js -s videojs-dock -d -o dist/tests.js",
     "watch-tests": "mkdir -p dist/ && watchify test/karma-qunit-shim.js test/*-test.js -u test/karma.conf.js -s videojs-dock -v -d -o dist/tests.js",
     "test": "jshint src/**.js test/**.js index.js && npm run build-tests && karma start test/karma.conf.js --single-run",
-    "prepublish": "npm run build",
+    "predist": "npm run babel",
+    "dist": "mkdir -p dist/ && browserify -t browserify-shim . -s videojs-dock -o dist/bundle.js",
```
Comment (gkatsev) [inline package.json]: The `dist` file needs to be browserified with `browserify-shim` as a transform. It's still configured below in this file.

File: package.json [hunk 4]
```diff
@@ -16,7 +16,10 @@
     "build-tests": "mkdir -p dist/ && browserify test/karma-qunit-shim.js test/*-test.js -u test/karma.conf.js -s videojs-dock -d -o dist/tests.js",
     "watch-tests": "mkdir -p dist/ && watchify test/karma-qunit-shim.js test/*-test.js -u test/karma.conf.js -s videojs-dock -v -d -o dist/tests.js",
     "test": "jshint src/**.js test/**.js index.js && npm run build-tests && karma start test/karma.conf.js --single-run",
-    "prepublish": "npm run build",
+    "predist": "npm run babel",
```
Comment (gkatsev) [inline package.json]: Make sure that the files are compiled via babel before browserify is run.

Comment (gkatsev) [inline package.json]: We can also use babelify and point browserify at `src/index.js`. Doing it this way means we can just `browserify .` and have it use the `main` field of this file.

File: package.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "name": "videojs-dock",
   "version": "1.0.3",
   "description": "Dock plugin for videojs",
-  "main": "index.js",
+  "main": "lib/index.js",
```
Comment (gkatsev) [inline package.json]: Point at the es5 code for browserify consumers of this package.

File: dist/bundle.js [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+
+function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }
+
+function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
+
+var videojs = (typeof window !== "undefined" ? window['videojs'] : typeof global !== "undefined" ? global['videojs'] : null);
```
Comment (gkatsev) [inline dist/bundle.js]: As you can see, this file looks for `videojs` in the global scope as opposed to still trying to require it.

File: lib/index.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }
+
+function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
+
+var videojs = require('video.js');
```
Comment (gkatsev) [inline lib/index.js]: we're still requiring `video.js` here.

**Comment by gkatsev** [comment]
@misteroneill @chemoish here's the proof of concept for creating a `dist/bundle.js` that looks for videojs globally but when used through require, expects to be able to require videojs.

**Comment by misteroneill** [comment]
Nice, this looks like a good path to head down. I'll definitely want to set up something based on this in the generator. :+1:

File: package.json [hunk 8]
```diff
@@ -58,12 +62,6 @@
     "video.js": "^5.0.0-rc.90",
```
Comment (gkatsev) [inline package.json]: also, this needs to move to dependencies with a non-RC semver-range (`^5.0.0`).

**Comment by chemoish** [comment]
Cool, I'll see how this information might fit in webpack. :+1:

**Comment by misteroneill** [comment]
Thinking further along this line... I don't feel like `lib/` should be named `lib/` or committed as it's really a build artifact rather than source. What about:

`dist/node/*.js` - Babel output, `"main"` entry point, for Node/Browserify consumers.
`dist/<plugin-name>.js` - shimmed/browser global.
`dist/<plugin-name>.min.js` - shimmed/browser global, minified.

**Comment by gkatsev** [comment]
Yep, lib shouldn't be committed to the repo.

**Comment by misteroneill** [comment]
Closing this due to the #5 merge. These issues are addressed by the generator conventions.

**Pull request closed without merging**